socs/agents/scpi_psu/drivers.py

```diff
@@ -1,6 +1,150 @@
 # Tucker Elleflot
+
+import socket
+import time
+
 from socs.common.prologix_interface import PrologixInterface
 
+# append new model strings as needed
+ONE_CHANNEL_MODELS = ['2280S-60-3', '2280S-32-6']
+THREE_CHANNEL_MODELS = ['2230G-30-1']
+
+
+class ScpiPsuInterface:
+    def __init__(self, ip_address, gpibAddr, port, **kwargs):
+        self.ip_address = ip_address
+        self.gpibAddr = gpibAddr
+        self.port = port
+        self.sock = None
+        self.model = None
+        self.numChannels = 0
+        self.conn_socket()
+        try:
+            self.configure()
+        except ValueError as err:
+            raise ValueError(err)
+        super().__init__(**kwargs)
+
+    def conn_socket(self):
+        self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
+        self.sock.connect((self.ip_address, self.port))
+        self.sock.settimeout(5)
+
+    def read(self):
+        return self.sock.recv(128).decode().strip()
+
+    def write(self, msg):
+        message = msg + '\n'
+        self.sock.sendall(message.encode())
+        time.sleep(0.1)  # to prevent flooding the connection
+
+    def identify(self):
+        self.write('*idn?')
+        return self.read()
+
+    def read_model(self):
+        idn_response = self.identify().split(',')[1]
+        if (idn_response.startswith('MODEL')):
+            return idn_response[6:]
+        else:
+            return idn_response
+
+    def configure(self):
+        self.model = self.read_model()
+        if (self.model in ONE_CHANNEL_MODELS):
+            self.numChannels = 1
+        if (self.model in THREE_CHANNEL_MODELS):
+            self.numChannels = 3
+        if (self.numChannels == 0):
+            raise ValueError('Model number not found in known device models', self.model)
+
+    def enable(self, ch):
+        '''
+        Enables output for channel (1,2,3) but does not turn it on.
+        Depending on state of power supply, it might need to be called
+        before the output is set.
+        '''
+        if (self.numChannels != 1):
+            self.set_chan(ch)
+            self.write('OUTP:ENAB ON')
+
+    def disable(self, ch):
+        '''
+        disabled output from a channel (1,2,3). once called, enable must be
+        called to turn on the channel again
+        '''
+        self.write('OUTP:ENAB OFF')
```

Comment:
Yes, at least the 2280S has a output enable/disable on the front panel that can also be controlled over SCPI. The documentation suggests that the single-channel device presumes the "CHAN" parameter is always "1".

Interestingly, it seems this function is never called for either of the drivers, nor is it referenced in the agent code, probably vestigial at this point. Should I remove it altogether (from both drivers)?

Reply:
Having extra functionality in the drivers that the agent doesn't take advantage of (yet) is fine. Sometimes users end up commanding things directly via the drivers for debugging, so I wouldn't remove it. But I'd drop the check on `self.numChannels != 1` in the `enable()` method, since that'll prevent it from enabling single channel devices if they're ever disabled.

Reply:
Ok, changed the `enable()` code to
```
if (self.num_channels != 1):
    self.set_chan(ch)
self.write('OUTP:ENAB ON')
```
which is probably what I was trying to do originally.